### PR TITLE
[Metrics] Metrics in core commissioning phase

### PR DESCRIPTION
**Pull request by anush-apple**
- Added metrics tracing in core parts of the commissionning flow:
  - PASE
  - OperationalSessionSetup
  - Different stages of commissioning

### CODE REVIEW

File: src/controller/CommissioningDelegate.cpp [hunk 1]
```diff
@@ -138,5 +138,124 @@ const char * StageToString(CommissioningStage stage)
...
+        return "core_cm_stg_secure_pairing";
+
+    case kReadCommissioningInfo:
+        return "core_cm_stg_rd_comm_info";
+
+    case kReadCommissioningInfo2:
+        return "core_cm_stg_rd_comm_info2";
+
+    case kArmFailsafe:
+        return "core_cm_stg_afs";
```
Comment (tcarmelveilleux) [inline src/controller/CommissioningDelegate.cpp:159]: Why these cryptic strings? Is there a limit to length?

File: src/app/OperationalSessionSetup.cpp [hunk 2]
```diff
@@ -242,6 +244,8 @@ void OperationalSessionSetup::UpdateDeviceData(const Transport::PeerAddress & ad
         return;
     }
 
+    MATTER_LOG_METRIC_END(kMetricDeviceOperationalDiscovery, CHIP_NO_ERROR);
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:247]: So how will this work in practice when we get multiple UpdateDeviceData for a single discovery start (e.g. if we found multiple IPs and the first one we tried did not work)?

File: src/app/OperationalSessionSetup.cpp [hunk 3]
```diff
@@ -591,6 +601,8 @@ CHIP_ERROR OperationalSessionSetup::LookupPeerAddress()
         return CHIP_NO_ERROR;
     }
 
+    MATTER_LOG_METRIC_BEGIN(kMetricDeviceOperationalDiscovery);
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:604]: Similar question here: it looks like we can reach this point multiple times before ever hitting a `MATTER_LOG_METRIC_END(kMetricDeviceOperationalDiscovery)`, no?

File: src/controller/AutoCommissioner.cpp [hunk 4]
```diff
@@ -24,12 +24,14 @@
 #include <controller/CHIPDeviceController.h>
 #include <credentials/CHIPCert.h>
 #include <lib/support/SafeInt.h>
+#include <tracing/metric_event.h>
```
Comment (bzbarsky-apple) [inline src/controller/AutoCommissioner.cpp:27]: Why the header and namespace change in this file?  I don't see any metrics being recorded in this file....

File: src/controller/CHIPDeviceController.cpp [hunk 5]
```diff
@@ -736,6 +744,7 @@ CHIP_ERROR DeviceCommissioner::EstablishPASEConnection(NodeId remoteDeviceId, Re
                     // We already have an open secure session to this device, call the callback immediately and early return.
                     mPairingDelegate->OnPairingComplete(CHIP_NO_ERROR);
                 }
+                MATTER_LOG_METRIC_END(kMetricDeviceCommissionerPASESession, CHIP_NO_ERROR);
                 return CHIP_NO_ERROR;
             }
             if (current->IsSessionSetupInProgress())
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:750]: Do we need an end metric inside this block too?

Or perhaps: should we only have the "start" metric when we get around to actually doing the `Pair()` call?  I don't quite understand what the `kMetricDeviceCommissionerPASESession` represents given the code here right now.

File: src/controller/CHIPDeviceController.cpp [hunk 6]
```diff
@@ -908,6 +920,7 @@ CHIP_ERROR DeviceCommissioner::Commission(NodeId remoteDeviceId)
     mDefaultCommissioner->SetOperationalCredentialsDelegate(mOperationalCredentialsDelegate);
     if (device->IsSecureConnected())
     {
+        MATTER_LOG_METRIC_BEGIN(kMetricDeviceCommissionerCommission);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:923]: So is kMetricDeviceCommissionerCommission a metric that gets begin/end?  How do the `ReturnErrorOnFailureWithMetric`/`VerifyOrDoWithMetric` bits interact with begin/end?

File: src/controller/CHIPDeviceController.cpp [hunk 7]
```diff
@@ -1960,6 +1982,7 @@ void DeviceCommissioner::OnDeviceConnectedFn(void * context, Messaging::Exchange
 void DeviceCommissioner::OnDeviceConnectionFailureFn(void * context, const ScopedNodeId & peerId, CHIP_ERROR error)
 {
     // CASE session establishment failed.
+    MATTER_LOG_METRIC_END(kMetricDeviceOperationalSetup, error);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1985]: This seems like a very generic metric name, for a place which is not involved in most "operational setup". This bit is specific to operational session setup at the end of commissioning.

File: src/controller/SetUpCodePairer.cpp [hunk 8]
```diff
@@ -89,10 +93,16 @@ CHIP_ERROR SetUpCodePairer::PairDevice(NodeId remoteId, const char * setUpCode,
...
 
-    ReturnErrorOnFailure(Connect(payload));
-    return mSystemLayer->StartTimer(System::Clock::Milliseconds32(kDeviceDiscoveredTimeout), OnDeviceDiscoveredTimeoutCallback,
-                                    this);
+    ReturnErrorOnFailureWithMetric(kMetricSetupCodePairerPairDevice, Connect(payload));
+    auto errorCode =
+        mSystemLayer->StartTimer(System::Clock::Milliseconds32(kDeviceDiscoveredTimeout), OnDeviceDiscoveredTimeoutCallback, this);
+    if (CHIP_NO_ERROR == errorCode)
+    {
+        MATTER_LOG_METRIC_BEGIN(kMetricSetupCodePairerPairDevice);
```
Comment (bzbarsky-apple) [inline src/controller/SetUpCodePairer.cpp:101]: Again, I don't quite understand how MATTER_LOG_METRIC_BEGIN and ReturnErrorOnFailureWithMetric work when using the same metric name....

File: src/protocols/secure_channel/CASESession.cpp [hunk 9]
```diff
@@ -47,6 +47,7 @@
 #include <system/SystemClock.h>
 #include <system/TLVPacketBufferBackingStore.h>
 #include <tracing/macros.h>
+#include <tracing/metric_event.h>
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.cpp:50]: Where is this used in here?

File: src/controller/CHIPDeviceController.cpp (hunk 5 above)
Comment (anush-apple) [reply src/controller/CHIPDeviceController.cpp:750]: Actually I don't think we want to end here. As the comment says, the delegate will be called and that is where the end of the PASE session will be marked

File: src/controller/CHIPDeviceController.cpp (hunk 6 above)
Comment (anush-apple) [reply src/controller/CHIPDeviceController.cpp:923]: The ReturnErrorOnFailureWithMetric/VerifyOrDoWithMetric are used to handle any failure that happen synchronously before the log async operation to start commissioning.

File: src/controller/SetUpCodePairer.cpp (hunk 8 above)
Comment (anush-apple) [reply src/controller/SetUpCodePairer.cpp:101]: The begin and end are used to make longer sync stages. The ReturnErrorOnFailureWithMetric is used for any errors that happen before the begin is marked. If there are errors, they are just marked with the same even name but not begin/end modes

File: src/app/OperationalSessionSetup.cpp (hunk 3 above)
Comment (anush-apple) [reply src/app/OperationalSessionSetup.cpp:607]: Yes, that is fine. Backend needs to handle this

File: src/protocols/secure_channel/CASESession.cpp (hunk 9 above)
Comment (woody-apple) [reply src/protocols/secure_channel/CASESession.cpp]: Resolved by my eye

File: src/controller/SetUpCodePairer.cpp (hunk 8 above)
Comment (woody-apple) [reply src/controller/SetUpCodePairer.cpp:101]: Yup

File: src/controller/CHIPDeviceController.cpp (hunk 6 above)
Comment (woody-apple) [reply src/controller/CHIPDeviceController.cpp:923]: Yup

File: src/controller/CHIPDeviceController.cpp (hunk 5 above)
Comment (woody-apple) [reply src/controller/CHIPDeviceController.cpp:750]: Agree

File: src/app/OperationalSessionSetup.cpp (hunk 3 above)
Comment (woody-apple) [reply src/app/OperationalSessionSetup.cpp:607]: Yup

File: src/app/OperationalSessionSetup.cpp [hunk 10]
```diff
@@ -305,9 +307,14 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
     mCASEClient = mClientPool->Allocate();
     ReturnErrorCodeIf(mCASEClient == nullptr, CHIP_ERROR_NO_MEMORY);
 
+    // Session has been established. This is when discovery is also stopped
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:310]: I don't get this comment.  The session has not been established at this point....

And as we discussed discovery is sill ongoing, in parallel with the CASE bits.  So I'm a little confused by this part.

File: src/app/OperationalSessionSetup.cpp [hunk 11]
```diff
@@ -605,6 +616,10 @@ CHIP_ERROR OperationalSessionSetup::LookupPeerAddress()
         return CHIP_NO_ERROR;
     }
 
+    // This call be called multiple times for each discovered address.The metric backend can handle this 
+    // and always pick the earliest occurance as the start of the event.
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:620]:
```suggestion
    // This code can be reached multiple times, if we discover multiple addresses or do retries.
    // The metric backend can handle this and always picks the earliest occurrence as the start of the event.
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp [hunk 12]
```diff
@@ -666,16 +668,20 @@ CHIP_ERROR DeviceCommissioner::PairDevice(NodeId remoteDeviceId, const char * se
...
 
 CHIP_ERROR DeviceCommissioner::PairDevice(NodeId remoteDeviceId, RendezvousParameters & rendezvousParams,
                                           CommissioningParameters & commissioningParams)
 {
     MATTER_TRACE_SCOPE("PairDevice", "DeviceCommissioner");
-    ReturnErrorOnFailure(EstablishPASEConnection(remoteDeviceId, rendezvousParams));
-    return Commission(remoteDeviceId, commissioningParams);
+    ReturnErrorOnFailureWithMetric(kMetricDeviceCommissionerCommission, EstablishPASEConnection(remoteDeviceId, rendezvousParams));
+    auto errorCode = Commission(remoteDeviceId, commissioningParams);
+    VerifyOrDoWithMetric(kMetricDeviceCommissionerCommission, CHIP_NO_ERROR == errorCode, errorCode);
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:683]: Sounds like you need a `VerifySuccessOrDoWithMetric` for the common case of checking errorCode == CHIP_NO_ERROR.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp (hunk 12 above)
Comment (anush-apple) [reply src/controller/CHIPDeviceController.cpp:683]: Thanks for the suggestion. I can add that next

Comment (anush-apple) [reply src/controller/CHIPDeviceController.cpp:683]: Filed https://github.com/project-chip/connectedhomeip/issues/33508

**Pull request merged**